src/cider/nrepl/middleware/util/error_handling.clj

```diff
@@ -0,0 +1,137 @@
+    return a map that will form the basis of the nREPL reply message.
+  - coll => The coll will be turned into a set which is used as the
+    reply message's status.
+  - kw => Wrapped in a set and used as the reply message's status.
+  - ::default => used as the default error handler, which simply adds
+    a reasonably named keyword (ie, `:<op>-error`) to the status."
+  [input & _]
+  (cond (= ::default input) :default
+        (fn? input) :function
+        (map? input) :inline-reply
+        (coll? input)  :status-coll
+        (keyword? input) :status-item))
+
+;;; ERROR HANDLER - see selector docstring
+
+(defmulti error-handler selector)
+
+(defmethod error-handler :inline-reply [answer msg e]
+  (let [reply           (base-error-response msg e)
+        terminal-status (set/union #{:done} (normalize-status (:status answer)))]
+    (response-for msg (assoc reply :status terminal-status))))
+
+(defmethod error-handler :function [f msg e]
+  (error-handler (f msg e) msg e))
+
+(defmethod error-handler :status-coll [statuses msg e]
+  (error-handler {:status (set statuses)} msg e))
+
+(defmethod error-handler :status-item [status msg e]
+  (error-handler {:status (set [status])} msg e))
+
+(defmethod error-handler :default [_ msg e]
+  (error-handler (keyword (str (:op msg) "-error")) msg e))
+
+;;; OP HANDLER - see selector docstring
+
+(defmulti op-handler selector)
+
+(defmethod op-handler :inline-reply [answer msg]
+  (let [terminal-status (set/union #{:done} (normalize-status (:status answer)))]
+    (response-for msg (assoc answer :status terminal-status))))
+
+(defmethod op-handler :function [f msg]
+  (op-handler (f msg) msg))
+
+(defmethod op-handler :status-coll [statuses msg]
+  (op-handler {:status (set statuses)} msg))
+
+(defmethod op-handler :status-item [status msg]
+  (op-handler {:status (set [status])} msg))
+
+;;; SAFE TRANSPORT WRAPPER
+
+(defmacro with-safe-transport
+  "This will safely handle all the transport calls mapped out in the
+  `wrap-<middleware>` functions. All checked exceptions will be
+  caught, analyzed by the `cider.nrepl.middleware.stacktrace`
+  middleware, and an error message will be returned to the client with
+  a stacktrace renderable by the default CIDER stacktrace viewer.
+  Takes the default pass-through handler as well as a list of pairings
+  between op names and actions used to process the ops as
+  varargs. Actions can either be expressed as a 2-item vector with the
+  head being the op-action and the tail being the error-action, or if
+  the default error handler is sufficient, then the op name can be
+  paired directly to the op-action.
+
+  Actions can be functions, maps, non-associate collections, and
+  single items such as kw's, strings, numbers, etc. The utilization of
+  each type is discussed above in the `selector` method."
+  [pass-through & pairings]
+  `(fn [{op# :op transport# :transport :as msg#}]
+     (try
```

Comment:
Oh wow I didn't even think about how ugly this would look once macroexpanded. The outer `try` is needed so you can do the inline responses safely (say for something that has a constant reply), but we'll probably never use this and could remove it entirely.